### PR TITLE
Fixed absolute path issues to allow use of virtual host monster and proxies...

### PR DESCRIPTION
**Pull request by gutow**
I have worked my way through the notebook templates and updated links to be mostly absolute and be virtual host monster aware so that the sage notebook can be properly proxied behind Apache2 or nginx.  It is now possible to run multiple sage server instances on one domain name.  I think I have found all the things that were broken, but this does need testing by others.  This does not change local usage although the path names are slightly different.

### CODE REVIEW

**Comment by kcrisman** [comment]
Thanks for all this.  I will try to look at this but there are quite a few changes (if many relatively minor/straightforward) and it will take me some time to look at them, with classes beginning in just a few days.

My major concern, though, is backward incompatibility - for instance, do publish urls change?  I'm wondering whether any other urls that one might have otherwise had in a link somewhere would change with this.  You are right that presumably this will help with some of the scripting vulnerability.

**Comment by gutow** [comment]
Good point on the published worksheets.  I will double check, but I believe that is taken care of internally on the server.  I meant it to.  However, this will not fix external links from other web sites.  One possibility is to warn people maintaining a older server to add a redirect from the root of their domain to the new site name.

**Comment by kcrisman** [comment]
> However, this will not fix external links from other web sites. 

Do you mean that if someone published a worksheet at e.g. http://my.sage.server.edu/home/pub/123 then it would no longer work from links in published articles, web links, etc?  That is not really good, especially considering how many broken links already exist because of sagenb.org no longer having published worksheets...

**Comment by gutow** [comment]
>Do you mean that if someone published a worksheet at e.g. http://my.sage.server.edu/home/pub/123 then it would no longer work from links in published articles, web links, etc?

Yes, but if you want to maintain compatibility on your server with old external links all you have to do is add the following rewrite rule to the end of your rewrite chain.  The only problem with it is that you then cannot serve anything else (say plain vanilla html) at the root of your site.

        RewriteRule ^/(*.) http://my.sage.server.edu/new_site_name$1 [R]

I might be wrong about the ``[R]``, for 'redirect', but if this needs to be made clear I can verify it with the Apache2 documentation before adding it to ours.

**Comment by kcrisman** [comment]
It looks like in most cases you basically have added the `url_root` to all paths.  Interesting.  I don't quite understand why that is necessary, but as I said I know nothing about hosting.

More mundanely, I note that you have two apparently identical sets of commits up to the merge https://github.com/gutow/sagenb_slickgrid/commit/90f088f7ee10965e2ffa7430abd6102512fe80f6. Is there something different about those two sets that should be looked at closely?

**Comment by gutow** [comment]
It does look like the two commits are identical.  There are probably slight differences in spacing or some quotations as I was having trouble with the REST formatting.  As it does mess up the final product I would not worry about it.

**Comment by gutow** [comment]
Is there some other place where documentation should be changed (maybe the readme for sagenb)?

**Comment by kcrisman** [comment]
> Is there some other place where documentation should be changed (maybe the readme for sagenb)?

Probably, but as I said this is a more major change whose implications I haven't fully thought through yet, so you have time to think about that.  On the other hand, we don't really document most of the paths, presumably purposely - just links.  My main concern would be published worksheets and making sure we caught everything - there must be hundreds.  You could check out whether the translations all work properly.

**Comment by kcrisman** [comment]
Also, I'm wondering whether there is something more robust than `url_root+"?typ=trash"` or if that is just how it must be done as javascript has no access to Python things like `os.sep`.

**Comment by gutow** [comment]
>Also, I'm wondering whether there is something more robust than `url_root+"?typ=trash"` or if that is just how it must be done as javascript has no access to Python things like `os.sep`.

Technically `url_root` could be equal to the  "/" that it replaced in that particular js line or any other path that I have fixed in the server code and templates.  `os.sep` is appropriate to take care of things that vary between operating systems.  For url's there is no alternative but a "/" separated path.  I actually think that this is more robust than the hardcoded root = "/" that we were using previously. Now it can be set at server launch time and we are putting absolute paths in our pages rather then relative.

**Comment by gutow** [comment]
> we caught everything - there must be hundreds
Not really.  Except for the url paths for actions the paths to worksheets are all taken care of in a single place.  I did not have to touch that code.  Just the code that generates the url for the link to a worksheet.

**Comment by kcrisman** [comment]
>  os.sep is appropriate to take care of things that vary between operating systems. For url's there is no alternative but a "/" separated path. 

Oh, of course, yes that makes sense.

**Comment by kcrisman** [comment]
One idea I had about the published worksheets is that perhaps one could just have `http://my.sage.server.edu/home/pub/123` automatically redirect to `http://my.sage.server.edu/home/sage/pub/123` or whatever was appropriate in the Sage server logic.  So that it would just detect this.  I don't want to assume everyone is using any given server and with a rewrite - for instance, on my local computer I am presumably running sagenb using twisted!

**Comment by gutow** [comment]
NOTE: because I need this to work I am pushing the small fix to the import paths for preparsing to maintain campatibility with Sage-6.4.  This will have to be changed back for Sage-6.5 compatibility.
The patch that needs reverting is copied below:

        --- /home/jonathan/git/sagenb-11.2/sagenb/misc/support.py
        +++ /home/jonathan/git/sagenb_slickgrid/sagenb/misc/support.py
        @@ -492,13 +492,13 @@
         # Preparser
         ###################################################
         try:
        -    from sage.repl.preparse import preparse, preparse_file
        +    from sage.misc.preparser import preparse, preparse_file
             def do_preparse():
                 """
                 Return True if the preparser is set to on, and False otherwise.
                 """
        -        import sage.repl.interpreter
        -        return sage.repl.interpreter._do_preparse
        +        import sage.misc.interpreter
        +        return sage.misc.interpreter.do_preparse
         except ImportError:
             def preparse(line, *args, **kwds):
                 return line
        @@ -666,7 +666,7 @@
             Preparse the contents of a worksheet cell in the notebook,
             respecting the user using ``preparser(False)`` to turn off the
             preparser.  This function calls
        -    :func:`~sage.repl.preparse.preparse_file` which also reloads
        +    :func:`~sage.misc.preparse.preparse_file` which also reloads
             attached files.  It also does displayhook formatting by calling
             the :func:`~sagenb.notebook.interfaces.format.displayhook_hack`
             function.
        @@ -676,7 +676,7 @@
             - ``s`` - a string containing code
     
             - ``globals`` - a string:object dictionary; passed directly to
        -      :func:`~sage.repl.preparse.preparse_file`
        +      :func:`~sage.misc.preparse.preparse_file`
 
             OUTPUT:

**Comment by kcrisman** [comment]
That makes eminent sense, and should solve the problem you have, for sure.

**Comment by gutow** [comment]
This is once again ready for review and inclusion in the notebook.  I am using this on my production class server.  A test server running this is available at https://141.233.196.149/sagetest/.
To illustrate why this is useful my production server is on the same machine at https://141.233.196.149/sage/.  I'm also running a test joomla server at the same address.

**Comment by kcrisman** [comment]
> To illustrate why this is useful 

I might be just too woefully ignorant, but what is the difference between them, other than that there is no way to create accounts on the latter server?

**Comment by kcrisman** [comment]
I do get this when plotting `sphere()`

    /home/sage8888/sage-6.8/local/lib/python2.7/site-packages/setuptools-12.\
4-py2.7.egg/pkg_resources/__init__.py:1224: UserWarning:
/home/sage8888/.sage//.python-eggs is writable by group/others and
vulnerable to attack when used with get_resource_filename. Consider a
more secure location (set with .set_extraction_path or the
PYTHON_EGG_CACHE environment variable).

but I think that others have seen this as well.

**Comment by gutow** [comment]
My two servers are presently the same, although only people I give accounts to can access the second one.  I use that for graded class assignments.  The first one is where I try things.  Often it is a bleeding edge version of Sage or the SageNotebook.  Soon I will probably try it with the NewUI interface as I really want the codemirror functionality.

As to the funny message, it appears to be something about the way permissions are set in the compiled version supplied for Linux.  I do not understand what it relates to.  It may be the way the permissions are set in repository where the notebook data is stored.  I haven't been able to trace it yet.  If shows up in all versions of 6.8 I've tried, so is unrelated to this pull request.

**Comment by kcrisman** [comment]
>  If shows up in all versions of 6.8 I've tried, so is unrelated to this pull request.

That's what I figured.

**Comment by kcrisman** [comment]
And I assume that the published worksheet URL problem is still there, correct?

**Comment by gutow** [comment]
Do you mean if people have previously published a worksheet using a server living at the root of their server outside links will not work?  If proper redirects are not done on the server that is correct.  However it would be easy enough to put in a rewrite rule along the lines of:

RewriteRule ^/pub(.*) https:XXX.YYY.ZZZ:433/<sage_site_name>/pub$1 [R]

This would redirect everything going to the /pub directory to the new /pub directory.

**Comment by kcrisman** [comment]
Right, you said that, but wouldn't that (technically speaking) be outside the aegis of sagenb?

(As a related question, I seem to recall that those URL changes are whether you use proxying or not, right? I knew this once but haven't looked at the code for some time.)

Also, I note that you have XXX.YYY etc. - but one could also do something along the lines of sage.XYZ.edu still, right?  Also, why 433 - isn't sagenb usually on port 8080, or is that again betraying my ignorance of how public websites must function?  (I assume the latter.)

**Comment by gutow** [comment]
OK, I'm obviously not explaining myself clearly enough, which suggest this pull request may need additional specific instructions (wiki additions maybe, or part of the notebook/sage readme?).  Let me try to be clearer:
A) This change means that a sage notebook will never directly reside at a web address of the form http://(or https://)domain.name.extension/.  The default address will be http://domain.name.extension/sage/ (or for running locally http://localhost:8080/sage/).
B) The https://domain.name.extension is, of course, a secure http address.  The :443 (I see my sticky keyboard gave me 433. Ah, that's why I've been having trouble with my password at work!) is the default port for https communications.  I've found that including it explicitly in redirects avoids any flakiness in browser address parsing.
C) If one were setting up a publicly accessible server, it should be served behind an industrial strength server such as Apache2 (my favorite) or nginx (said to be less resource intensive and thus faster).
  1. To put sage behind a server you have sage only respond on a local port (e.g. localhost:808X, where X is a digit of your choice).  Run the notebook command as follows.
     * notebook(interface='localhost', port=8081, automatic_login=False, timeout=1200, site_name="your_chosen_site_name")
  2. Since people's accounts are password protected outside connections should be made using https protocols.  So the server should be setup to handle https (this is server dependent and explained elsewhere).
  3. The outside requests for the sage notebook need to be passed through the server to the local instance.  In Apache2 this is done using rewrite rules (for details see the documentation on mod_rewrite.  There are good instructions for activating this in the Apache2 documentation.).  In your site configuration file you then need a line like.
     * RewriteRule ^/your_chosen_site_name(.*) http://localhost:8081/virtualhostbase/https/domain.name.extension/your_chosen_site_name$1 [P]
  4. If you are updating an instance that was previously being served directly at the site root and you want old links to /pub to work you will need an additional rewrite rule prior to the above.  This rule just redirects the request to the proper address.
   * RewriteRule ^/pub(.*) https://domain.name.extension:433/your_chosen_site_name/pub$1 [R]
  5. Why do we want this?
    * The industrial servers are more secure and handle encryption more efficiently.
    * This allows multiple servers to be run on the same machine.  My primary reason for doing this is to test new updates before updating my production server.  A secondary reason is that I have limited computer resources and need to run different web applications on the same computer with a single IP address.

**Comment by kcrisman** [comment]
>
> locally http://localhost:8080/sage/).
>
>
> Thanks, most of that is of course my ignorance.  However, the above
probably requires a significant change in some documentation in quite a few
places that say just to go to "localhost:8080", among others the virtual
machine implementation for Windows.

Is it possible to have this all be a *switch* that could be turned on or
off?  I know that is asking something big, potentially, so I'm only
wondering about the feasibility at this point.

**Comment by gutow** [comment]
A switch might be possible.  I see two problems with a switch:
1) We should not encourage anything that makes people use the notebook publicly without an industrial server in front for security reasons.
2) I will have to think about it further, but my first impression is that a switch would require quite a bit more code and significant additional logic in the page templates.

For local usage this pull should not matter as the automatic login will point them to the right place.  The notebook startup message is also fixed in this pull request to point to the correct address.

**Comment by gutow** [comment]
* One possibility is to split the notebook into two releases (legacy and ?) and encourage people to shift to this version.  
* I also worry about putting too much effort into this whole thing if we are going to get a reasonable SMC that can be launched locally.  I'm only still working on this because I need something that works now both locally and on a server.

**Comment by kcrisman** [comment]
By the way, is http://stackoverflow.com/questions/32479041/serve-web-pages-from-directory-other-than-var-www/32484178#32484178 conceivably related to this issue?  Just wondering.

**Comment by kcrisman** [comment]
Incidentally, this has merge conflicts - I'm sure they're trivial.  If we can get SOME method to fix the `pub` directory, even if it's very very explicit instructions for the above, though that is much less optimal than e.g. an "internal rewrite rule" that would make `/pub/` turn into `/sage/pub`.  I think that the other ones probably won't be as much of an issue, as long as all the links still work relative to each other, which I assume is the case if you've been using this without problems for quite a while :smile:

**Comment by gutow** [comment]
No can do the internal pub redirection as the outward facing server will not pass /pub to the notebook server without an explicit redirection at the outward facing server point.  How would you like the instructions more explicit?

Jonathan
> On Nov 12, 2015, at 10:27 AM, kcrisman <notifications@github.com> wrote:
> 
> Incidentally, this has merge conflicts - I'm sure they're trivial. If we can get SOME method to fix the pub directory, even if it's very very explicit instructions for the above, though that is much less optimal than e.g. an "internal rewrite rule" that would make /pub/ turn into /sage/pub. I think that the other ones probably won't be as much of an issue, as long as all the links still work relative to each other, which I assume is the case if you've been using this without problems for quite a while 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/sagemath/sagenb/pull/328#issuecomment-156155932>.
> 

                        Dr. Jonathan H. Gutow
Chemistry Department                                 gutow@uwosh.edu <mailto:gutow@uwosh.edu>
UW-Oshkosh                                                Office:920-424-1326
800 Algoma Boulevard                                 FAX:920-424-2042
Oshkosh, WI 54901
                http://www.uwosh.edu/facstaff/gutow/ <http://www.uwosh.edu/facstaff/gutow/>

**Comment by kcrisman** [comment]
> No can do the internal pub redirection as the outward facing server will
> not pass /pub to the notebook server without an explicit redirection at the
> outward facing server point. How would you like the instructions more
> explicit?
>
>
>
I mean that the instructions would take the form of a pull request to the
documentation of sagenb, and be as explicit as possible to say which file
they go in, which syntax, etc.  I don't know if that's possible, though
your comments earlier make it sound like it is.

**Comment by kcrisman** [comment]
@gutow - I'd like to finally look at this, thanks so much for your patience while a lot of other stuff was happening.  I will ask for two things:
* Fix merge conflict - probably fairly trivial
* Add instructions in some appropriate place - possibly including the README.rst file - with some rewrite rules that would work for both the situation where one proxies *and* the situation where one decides not to proxy.  I know very little about that and would be hesitant to say anything :frowning:

File: sagenb/data/sage/html/history.html [hunk 1]
```diff
@@ -3,13 +3,13 @@
...
 
 {% block javascript %}
-<script type="text/javascript" src="/javascript/sage/js/ws_list.js"></script>
+<script type="text/javascript" src="{{request.url_root}}/javascript/sage/js/ws_list.js"></script>
 {% endblock %}
 
 {% block page_id %}history-page{% endblock %}
 
 {% block body %}
-    <pre>{{ text | escape }}</pre>
```
Comment (kcrisman) [inline sagenb/data/sage/html/history.html]: Why remove the escape?  I feel like that is from some other ticket.

Comment (gutow) [inline sagenb/data/sage/html/history.html]: There were a bunch of these and other alternatives that were causing page rendering errors.  I no longer remember the detailed problems, but in the first commit where I started doing this there should be a comment.  I will have to dig.

Comment (kcrisman) [inline sagenb/data/sage/html/history.html]: I think that @migeruhito actually pointed out that you merged some, but not all, of his stuff in #324 and that is what happened.

**Comment by kcrisman** [comment]
Okay @gutow just letting you know that I've merged everything I'm planning to for now, with exception of #351 but that shouldn't affect your branch.  No rush of course, just fyi.

**Comment by dimpase** [comment]
Could you update me on the state of affairs on this pull request? Is it mere rebasing to be done? Thanks.

**Comment by gutow** [comment]
I suspect it just needs to be rebased.  I have not had time to look at it recently.  My site which makes use of it it still running on 6.8.  I think it got held up because people were worried about the fact that it makes it so that Sage does not occupy the root address of the server.  The only downside I see to that is that it is a change.  It makes it much easier to run multiple types of sites on one machine.  This is a big boon for testing and people like me who maintain a lot of low use (1 class) sites on a single scrounged server, for which I cannot get more than one IP and for which I have limited control over which ports can be open because I sit behind my Campus firewall.

Anyway, I can probably answer questions, but do not presently have time to do the fine work of chasing conflicts.

**Comment by gutow** [comment]
As sagenb is not really supported anymore, I am closing this pull request.

**Comment by kcrisman** [comment]
>
>
> As sagenb is not really supported anymore,
>

More's the pity.


> I am closing this pull request.
>
Thank you so much for all your support over the years!!!  Both in code and
moral.

**Pull request closed without merging**